Version 4

fetzi opened the pull request:
fixes #23, #31, #32

Commit:
Add support for magic properties and root key

fixes #23, #31, #32

Commit:
Fix test command for github action

Commit:
Add information about transformer generation

Witti commented on src/ClassBindings.php:
```diff
@@ -101,13 +69,22 @@ public function register($binding)
     }
 
-    private function handleRaw($property, $data, $propertyAccessor)
+    private function handleBinding(Binding $binding, Property $property, $data)
```
Docblocks missing

Witti commented on src/ClassBindings.php:
```diff
@@ -33,47 +34,14 @@ public function __construct(JsonDecoder $jsonDecoder)
      */
     public function decode($data, $instance)
```
type for $data is missing

Witti commented on src/JsonDecoder.php:
```diff
@@ -111,4 +139,94 @@ protected function transformRaw($jsonArrayData, $instance)
+        $jsonData = json_decode($json, true);
+
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            throw new InvalidJsonException();
```
Please add InvalidJsonException to the docblock

Witti commented on src/JsonDecoder.php:
```diff
@@ -75,17 +113,7 @@ public function decodeArray($jsonArrayData, $classType)
-    }
-
-    private function transform($transformer, $jsonArrayData, $instance)
+    protected function transform($transformer, $jsonArrayData, $instance)
```
Docblock missing

Witti commented on src/Property.php:
```diff
@@ -0,0 +1,59 @@
+     */
+    private $property;
+
+    public static function create($instance, string $propertyName)
```
Docblocks missing

Commit:
Fix bug in validate DateTimeBinding

Commit:
Apply requested changes

Commit:
Add auto case feature

if enabled it will automatically try to generate the camel case vesion
of a field from snake or kebap case and if found will create an
AliasBinding for it

Fixes #34

Witti commented on src/Property.php:
```diff
@@ -0,0 +1,67 @@
+        $this->property = $property;
+    }
+
+    public function set($value)
```
Docblock missing

Witti commented on src/Property.php:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    public function getName()
```
Docblock missing

Witti commented on src/JsonDecoder.php:
```diff
@@ -105,10 +169,98 @@ protected function transformRaw($jsonArrayData, $instance)
+     *
+     * @param string $class the class to scan
+     * @return array        the list of generated bindings
+     */
```
`@throws ReflectionException` missing

Witti commented on src/ClassBindings.php:
```diff
@@ -101,13 +82,80 @@ public function register($binding)
+        (new RawBinding($property->getName()))->bind($this->jsonDecoder, $data, $property);
+    }
+
+    private function autoCase(string $jsonField, $instance): ?Property
```
Docblock missing

Witti commented on src/ClassBindings.php:
```diff
@@ -101,13 +82,80 @@ public function register($binding)
+        return null;
+    }
+
+    private function snakeToCamelCase(string $input)
```
Docblock missing

Witti commented on src/ClassBindings.php:
```diff
@@ -101,13 +82,80 @@ public function register($binding)
+        return preg_replace_callback('/_([a-z])/', $fn, strtolower($input));
+    }
+
+    private function kebapToCamelCase(string $input)
```
Docblock missing

Witti reviewed (changes requested):
Sorry, found some missing docblocks again 😁

Commit:
Add missing doc blocks
